lib/at_cmd_parser/at_cmd_parser.c

```diff
@@ -0,0 +1,340 @@
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_valid_notification_char(char chr)
+{
+	chr = toupper(chr);
+
+	if ((chr >= 'A') && (chr <= 'Z')) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_terminated(char chr)
+{
+	if (chr == AT_CMD_BUFFER_TERMINATOR) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_separator(char chr)
+{
+	if ((chr == AT_PARAM_SEPARATOR) ||
+	    (chr == AT_CMD_SEPARATOR)) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_lfcr(char chr)
+{
+	if ((chr == '\r') ||
+	    (chr == '\n')) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_dblquote(char chr)
+{
+	if (chr == '"') {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_array_start(char chr)
+{
+	if (chr == '(') {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_array_stop(char chr)
+{
+	if (chr == ')') {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_number(char chr)
+{
+	if (isdigit(chr) ||
+	    (chr == '-') ||
+	    (chr == '+')) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline void set_new_state(enum at_parser_state new_state)
+{
+	if ((state != SEARCH_NEXT_PARAM) &&
+	    (state != DETECT_TYPE)) {
+		prev_state = state;
+	}
+
+	state     = new_state;
+}
+
+static inline void reset_state(void)
+{
+	prev_state = DETECT_TYPE;
+	state      = DETECT_TYPE;
+}
+
+/*
+ * Internal function.
+ * Parameters cannot be null. String must be null terminated.
+ */
+static int at_parse_param(const char ** at_params_str,
+			  struct at_param_list * const list,
+			  const size_t max_params)
+{
+	int index = 0;
+	const char *str = *at_params_str;
+
+	reset_state();
+
+	while ((!is_terminated(*str)) &&
+	       (index < max_params)) {
+
+		str += isspace(*str);
+
+		if (state == DETECT_TYPE) {
+			if ((index == 0) &&
+			    is_notification(*str)) {
+				/* Only first parameter in the string can be
+				   notification ID, (eg +CEREG:) */
+				set_new_state(NOTIFICATION);
+			} else if ((index > 0) &&
+				is_notification(*str)) {
+				/* If notifications is detected later in the
+				   string we should stop parsing and return
+				   EAGAIN */
+				break;
+			} else if (is_number(*str)) {
+				set_new_state(NUMBER);
+
+			} else if (is_dblquote(*str)) {
+				set_new_state(STRING);
+				str++;
+			} else if (is_array_start(*str)) {
+				set_new_state(ARRAY);
+				str++;
+			} else if (is_lfcr(*str) &&
+			          (prev_state == NUMBER)) {
+				/* If \n or \r is detected in the string and the
+				   previous param was a number we assume the
+				   next parameter is PDU data */
+
+				while (is_lfcr(*str)) {
+					str++;
+				}
+
+				set_new_state(SMS_PDU);
+			} else if ((index == 0) &&
+				    !is_notification(*str)) {
+				/* If the string start without an notification
+				   ID, we treat the whole string as one string
+				   parameter */
+				set_new_state(STRING);
+			} else {
+				if (is_separator(*str)) {
```

Comment:
Actually, this could be simply a next `else if (...)` statement, then you get the final `else` at the same level.

Reply:
Ok, let's try that and see if my tests agrees :)